### ISSUE TITLE
Object Storage Backup setup - Wasabi

### ISSUE THREAD

**Issue by tetricky** [longhorn/longhorn#2151]
I have tried to setup Wasabi (instead of aws) as a storage backend for longhorn backups. I appear to be unable to find a form that works.

I have referred to https://github.com/longhorn/longhorn/issues/811 which is a related issue, but I can't get where I need to be.

I have a secret, in the proscribed form (as per the docs)

apiVersion: v1
kind: Secret
metadata:
  name: private-name-secret
  namespace: longhorn-system
type: Opaque
data:
  AWS_ACCESS_KEY_ID: -base64 encoded access key- # access-key
  AWS_SECRET_ACCESS_KEY: -base64 encoded secret key- # secret-key
  AWS_ENDPOINTS: aHR0cHM6Ly9zMy5ldS1jZW50cmFsLTEud2FzYWJpc3lzLmNvbQ== # https://s3.eu-central-1.wasabisys.com

I've set "Backup Target Credential Secret" to "private-name-secret"

I've tried various forms of "Backup Target" and not been able to get anything working.

Is anyone successfully using wasabi s3 buckets?

Any pointers appreciated.

**Comment by joshimoo** [comment]
What are the forms of the backup target that you used?
Can you link to the Wasabi S3 setup instructions?

**Comment by tetricky** [comment]
The wasabi account user has two policy's attached. One is to restrict access to [bucket Name] and sub-directories. One is the longhorn policy (see below).

{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "GrantLonghornBackupstoreAccess0",
      "Effect": "Allow",
      "Action": [
        "s3:PutObject",
        "s3:GetObject",
        "s3:ListBucket",
        "s3:DeleteObject"
      ],
      "Resource": [
        "arn:aws:s3:::[Bucket Name]",
        "arn:aws:s3:::[Bucket Name]/*"
      ]
    }
  ]
}

Access to the bucket, using the users access and secret keys, and the endpoint https://s3.eu-central-1.wasabisys.com have been checked using minio client from the command line.

These details AWS_ACCESS_KEY_ID:(wasabi user access key); AWS_SECRET_ACCESS_KEY:(wasabi user secret key); AWS_ENDPOINTS: https://s3.eu-central-1.wasabisys.com have been base64 encoded and put added to a kubernetes secret. (see previous post)

In the longhorn settings (General) page "Backup Target Credential Secret" is set to the name of the secret with the encoded wasabi account settings (note, this is a paid for account, not a trial account). In the "Backup Target" a range of things have been tried. (eg, but not exclusively)

s3://[Bucket Name]@eu-central-1/longhorn/

error listing backups: error listing backup volumes: Failed to execute: /var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-v1.1.0/longhorn [backup ls --volume-only s3://[Bucket Name]@eu-central-1/longhorn/], output failed to list objects with param: { Bucket: "[Bucket Name]", Delimiter: "/", Prefix: "longhorn/" } response: { } error: AWS Error: RequestError send request failed Get "https://s3.eu-central-1.wasabisys.com/[Bucket Name]?delimiter=%!F(MISSING)&prefix=longhorn%!F(MISSING)": dial tcp: lookup s3.eu-central-1.wasabisys.com on 10.11.0.10:53: server misbehaving , stderr, time="2021-01-05T10:13:40Z" level=error msg="Fail to list s3: failed to list objects with param: {\n Bucket: \"[Bucket Name]\",\n Delimiter: \"/\",\n Prefix: \"longhorn/\"\n} response: {\n\n} error: AWS Error: RequestError send request failed Get \"https://s3.eu-central-1.wasabisys.com/[Bucket Name]?delimiter=%!F(MISSING)&prefix=longhorn%!F(MISSING)\": dial tcp: lookup s3.eu-central-1.wasabisys.com on 10.11.0.10:53: server misbehaving\n" pkg=s3 time="2021-01-05T10:13:40Z" level=error msg="failed to list objects with param: {\n Bucket: \"[Bucket Name]\",\n Delimiter: \"/\",\n Prefix: \"longhorn/\"\n} response: {\n\n} error: AWS Error: RequestError send request failed Get \"https://s3.eu-central-1.wasabisys.com/[Bucket Name]?delimiter=%!F(MISSING)&prefix=longhorn%!F(MISSING)\": dial tcp: lookup s3.eu-central-1.wasabisys.com on 10.11.0.10:53: server misbehaving\n" , error exit status 1

s3://[Bucket Name]/@eu-central-1/longhorn/

error listing backups: error listing backup volumes: Failed to execute: /var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-v1.1.0/longhorn [backup ls --volume-only s3://[Bucket Name]/@eu-central-1/longhorn/], output failed to list objects with param: { Bucket: "[Bucket Name]", Delimiter: "/", Prefix: "@eu-central-1/longhorn/" } response: { } error: AWS Error: MissingRegion could not find region configuration <nil> , stderr, time="2021-01-05T10:17:13Z" level=error msg="Fail to list s3: failed to list objects with param: {\n Bucket: \"[Bucket Name]\",\n Delimiter: \"/\",\n Prefix: \"@eu-central-1/longhorn/\"\n} response: {\n\n} error: AWS Error: MissingRegion could not find region configuration <nil>\n" pkg=s3 time="2021-01-05T10:17:13Z" level=error msg="failed to list objects with param: {\n Bucket: \"[Bucket Name]\",\n Delimiter: \"/\",\n Prefix: \"@eu-central-1/longhorn/\"\n} response: {\n\n} error: AWS Error: MissingRegion could not find region configuration <nil>\n" , error exit status 1

s3://[Bucket Name]/@eu-central-1/

error listing backups: error listing backup volumes: Failed to execute: /var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-v1.1.0/longhorn [backup ls --volume-only s3://[Bucket Name]/@eu-central-1/], output failed to list objects with param: { Bucket: "[Bucket Name]", Delimiter: "/", Prefix: "@eu-central-1/" } response: { } error: AWS Error: MissingRegion could not find region configuration <nil> , stderr, time="2021-01-05T10:19:03Z" level=error msg="Fail to list s3: failed to list objects with param: {\n Bucket: \"[Bucket Name]\",\n Delimiter: \"/\",\n Prefix: \"@eu-central-1/\"\n} response: {\n\n} error: AWS Error: MissingRegion could not find region configuration <nil>\n" pkg=s3 time="2021-01-05T10:19:03Z" level=error msg="failed to list objects with param: {\n Bucket: \"[Bucket Name]\",\n Delimiter: \"/\",\n Prefix: \"@eu-central-1/\"\n} response: {\n\n} error: AWS Error: MissingRegion could not find region configuration <nil>\n" , error exit status 1


s3://[Bucket Name]/@eu-central-1/

error listing backups: error listing backup volumes: Failed to execute: /var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-v1.1.0/longhorn [backup ls --volume-only s3://[Bucket Name]/@//], output failed to list objects with param: { Bucket: "[Bucket Name]", Delimiter: "/", Prefix: "@/" } response: { } error: AWS Error: MissingRegion could not find region configuration <nil> , stderr, time="2021-01-05T10:21:38Z" level=error msg="Fail to list s3: failed to list objects with param: {\n Bucket: \"[Bucket Name]\",\n Delimiter: \"/\",\n Prefix: \"@/\"\n} response: {\n\n} error: AWS Error: MissingRegion could not find region configuration <nil>\n" pkg=s3 time="2021-01-05T10:21:38Z" level=error msg="failed to list objects with param: {\n Bucket: \"[Bucket Name]\",\n Delimiter: \"/\",\n Prefix: \"@/\"\n} response: {\n\n} error: AWS Error: MissingRegion could not find region configuration <nil>\n" , error exit status 1

s3://[Bucket Name]/@s3.eu-central-1.wasabisys.com/longhorn/

error listing backups: error listing backup volumes: Failed to execute: /var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-v1.1.0/longhorn [backup ls --volume-only s3://[Bucket Name]/@s3.eu-central-1.wasabisys.com/longhorn/], output failed to list objects with param: { Bucket: "[Bucket Name]", Delimiter: "/", Prefix: "@s3.eu-central-1.wasabisys.com/longhorn/" } response: { } error: AWS Error: MissingRegion could not find region configuration <nil> , stderr, time="2021-01-05T10:25:54Z" level=error msg="Fail to list s3: failed to list objects with param: {\n Bucket: \"[Bucket Name]\",\n Delimiter: \"/\",\n Prefix: \"@s3.eu-central-1.wasabisys.com/longhorn/\"\n} response: {\n\n} error: AWS Error: MissingRegion could not find region configuration <nil>\n" pkg=s3 time="2021-01-05T10:25:54Z" level=error msg="failed to list objects with param: {\n Bucket: \"[Bucket Name]\",\n Delimiter: \"/\",\n Prefix: \"@s3.eu-central-1.wasabisys.com/longhorn/\"\n} response: {\n\n} error: AWS Error: MissingRegion could not find region configuration <nil>\n" , error exit status 1


s3://[Bucket Name]/@s3.eu-central-1.wasabisys.com/

error listing backups: error listing backup volumes: Failed to execute: /var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-v1.1.0/longhorn [backup ls --volume-only s3://[Bucket Name]/@s3.eu-central-1.wasabisys.com/], output failed to list objects with param: { Bucket: "[Bucket Name]", Delimiter: "/", Prefix: "@s3.eu-central-1.wasabisys.com/" } response: { } error: AWS Error: MissingRegion could not find region configuration <nil> , stderr, time="2021-01-05T10:28:07Z" level=error msg="Fail to list s3: failed to list objects with param: {\n Bucket: \"[Bucket Name]\",\n Delimiter: \"/\",\n Prefix: \"@s3.eu-central-1.wasabisys.com/\"\n} response: {\n\n} error: AWS Error: MissingRegion could not find region configuration <nil>\n" pkg=s3 time="2021-01-05T10:28:07Z" level=error msg="failed to list objects with param: {\n Bucket: \"[Bucket Name]\",\n Delimiter: \"/\",\n Prefix: \"@s3.eu-central-1.wasabisys.com/\"\n} response: {\n\n} error: AWS Error: MissingRegion could not find region configuration <nil>\n" , error exit status 1

As I say, this is not an exhaustive list...I've tried putting the endpoint in, or out, region in, or out...bucket in various places, with or without delimiter. I'm far from an expert, but I have tried to follow guidance and documentation, where I have bene able to find it.

Wasabi says it is s3 compatible.  I can stand that up with the other things that I do with it.  I can't configure longhorn to find it.

I know what the endpoint is: https://s3.eu-central-1.wasabisys.com
I know what the region is: eu-central-1 (although I don't believe it is required with that endpoint)
I have provided the access and secret keys, and the endpoint in the k8s-secret

I have set policies in wasabi.  I have confirmed access to the bucket. I have put a folder "longhorn" in the bucket, which ideally I would use - although I'd settle for anything working in the bucket at this point.

I have no further idea how to configure the longhorn settings to get it to connect.
Thank you for your quick response.

**Comment by joshimoo** [comment]
@tetricky can you try the below linked endpoint and url
https://github.com/longhorn/longhorn/issues/811#issuecomment-542058767

```
For the record I set the endpoint and target to:
endpoint: https://s3.wasabisys.com
backup target: s3://longhorn@eu-central-1/
```
Also ensure that there are no new lines as part of the access key / secret 
Which worked for the user in that thread.

There should not be an extra slash after `[Bucket Name]/`

**Comment by tetricky** [comment]
By way of research (and to see how each change affected the error reporting). I first just changed the backup target. "longhorn" is a folder in my bucket - which is  called [bucketname]. I first tried changing just the backup target, using the cuket name.

s3://[Bucket Name]@eu-central-1/

error listing backups: error listing backup volumes: Failed to execute: /var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-v1.1.0/longhorn [backup ls --volume-only s3://[Bucket Name]@eu-central-1/], output failed to list objects with param: { Bucket: "[Bucket Name]", Delimiter: "/", Prefix: "/" } response: { } error: AWS Error: RequestError send request failed Get "https://s3.eu-central-1.wasabisys.com/[Bucket Name]?delimiter=%!F(MISSING)&prefix=%!F(MISSING)": dial tcp: lookup s3.eu-central-1.wasabisys.com on 10.11.0.10:53: server misbehaving , stderr, time="2021-01-06T21:10:12Z" level=error msg="Fail to list s3: failed to list objects with param: {\n Bucket: \"[Bucket Name]\",\n Delimiter: \"/\",\n Prefix: \"/\"\n} response: {\n\n} error: AWS Error: RequestError send request failed Get \"https://s3.eu-central-1.wasabisys.com/[Bucket Name]?delimiter=%!F(MISSING)&prefix=%!F(MISSING)\": dial tcp: lookup s3.eu-central-1.wasabisys.com on 10.11.0.10:53: server misbehaving\n" pkg=s3 time="2021-01-06T21:10:12Z" level=error msg="failed to list objects with param: {\n Bucket: \"[Bucket Name]\",\n Delimiter: \"/\",\n Prefix: \"/\"\n} response: {\n\n} error: AWS Error: RequestError send request failed Get \"https://s3.eu-central-1.wasabisys.com/[Bucket Name]?delimiter=%!F(MISSING)&prefix=%!F(MISSING)\": dial tcp: lookup s3.eu-central-1.wasabisys.com on 10.11.0.10:53: server misbehaving\n" , error exit status 1

..and then using longhorn (I had zero expectation of theis being the right thing to do, but for completeness...). No useful information. Then I reverted to the bucket name in the backup target, and made sure the secret had no extra characters or lines.

s3://[Bucket Name]@eu-central-1/

error listing backups: error listing backup volumes: Failed to execute: /var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-v1.1.0/longhorn [backup ls --volume-only s3://[Bucket Name]@eu-central-1/], output failed to list objects with param: { Bucket: "[Bucket Name]", Delimiter: "/", Prefix: "/" } response: { } error: AWS Error: RequestError send request failed Get "https://s3.eu-central-1.wasabisys.com/[Bucket Name]?delimiter=%!F(MISSING)&prefix=%!F(MISSING)": dial tcp: lookup s3.eu-central-1.wasabisys.com on 10.11.0.10:53: server misbehaving , stderr, time="2021-01-06T21:21:05Z" level=error msg="Fail to list s3: failed to list objects with param: {\n Bucket: \"[Bucket Name]\",\n Delimiter: \"/\",\n Prefix: \"/\"\n} response: {\n\n} error: AWS Error: RequestError send request failed Get \"https://s3.eu-central-1.wasabisys.com/[Bucket Name]?delimiter=%!F(MISSING)&prefix=%!F(MISSING)\": dial tcp: lookup s3.eu-central-1.wasabisys.com on 10.11.0.10:53: server misbehaving\n" pkg=s3 time="2021-01-06T21:21:05Z" level=error msg="failed to list objects with param: {\n Bucket: \"[Bucket Name]\",\n Delimiter: \"/\",\n Prefix: \"/\"\n} response: {\n\n} error: AWS Error: RequestError send request failed Get \"https://s3.eu-central-1.wasabisys.com/[Bucket Name]?delimiter=%!F(MISSING)&prefix=%!F(MISSING)\": dial tcp: lookup s3.eu-central-1.wasabisys.com on 10.11.0.10:53: server misbehaving\n" , error exit status 1

Keeping the target the same, I then changed the endpoint, to the form without the eu-central-1 region (base64 encoded, in the secret, confirmed no extraneous characters).

"AWS_ENDPOINTS: aHR0cHM6Ly9zMy53YXNhYmlzeXMuY29t" - in secret.
s3://[Bucket Name]@eu-central-1/

error listing backups: error listing backup volumes: Failed to execute: /var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-v1.1.0/longhorn [backup ls --volume-only s3://[Bucket Name]@eu-central-1/], output failed to list objects with param: { Bucket: "[Bucket Name]", Delimiter: "/", Prefix: "/" } response: { } error: AWS Error: RequestError send request failed Get "https://s3.wasabisys.com/[Bucket Name]?delimiter=%!F(MISSING)&prefix=%!F(MISSING)": dial tcp: lookup s3.wasabisys.com on 10.11.0.10:53: server misbehaving , stderr, time="2021-01-06T21:28:04Z" level=error msg="Fail to list s3: failed to list objects with param: {\n Bucket: \"[Bucket Name]\",\n Delimiter: \"/\",\n Prefix: \"/\"\n} response: {\n\n} error: AWS Error: RequestError send request failed Get \"https://s3.wasabisys.com/[Bucket Name]?delimiter=%!F(MISSING)&prefix=%!F(MISSING)\": dial tcp: lookup s3.wasabisys.com on 10.11.0.10:53: server misbehaving\n" pkg=s3 time="2021-01-06T21:28:04Z" level=error msg="failed to list objects with param: {\n Bucket: \"[Bucket Name]\",\n Delimiter: \"/\",\n Prefix: \"/\"\n} response: {\n\n} error: AWS Error: RequestError send request failed Get \"https://s3.wasabisys.com/[Bucket Name]?delimiter=%!F(MISSING)&prefix=%!F(MISSING)\": dial tcp: lookup s3.wasabisys.com on 10.11.0.10:53: server misbehaving\n" , error exit status 1


So, as far as I can see, nothing doing.

**Comment by tetricky** [comment]
In the hope that it may assist.  I had a look at connecting from the same server (but not k8s - or in my case k3s) using rclone.

With a test file of "k3os-amd64.iso", in folder "longhorn" in bucket [BucketName]

[ws-server]
type = s3
provider = Wasabi
access_key_id = account_key
secret_access_key = account_secret
region = eu-central-1
endpoint = s3.eu-central-1.wasabisys.com
env_auth = false

rclone ls ws-server:[BucketName] # I can use with or without trailing delimter '/'
534796288 longhorn/k3os-amd64.iso

[ws-server]
type = s3
provider = Wasabi
access_key_id = account_key
secret_access_key = account_secret
region = eu-central-1
endpoint = s3.wasabisys.com
env_auth = false

rclone ls ws-server:[BucketName]

2021/01/07 20:10:36 ERROR : : error listing: AuthorizationHeaderMalformed: The authorization header is malformed; the region 'eu-central-1' is wrong; expecting 'us-east-1'
        status code: 400, request id: XXXXXXXXXXXXX, host id: YYYYYYYYYYYYY/ZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZ
2021/01/07 20:10:36 Failed to ls: AuthorizationHeaderMalformed: The authorization header is malformed; the region 'eu-central-1' is wrong; expecting 'us-east-1'
        status code: 400, request id: XXXXXXXXXXXXX, host id: YYYYYYYYYYYYY/ZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZ

So I would conclude that an endpoint s3.wasabisys.com with a region set to eu-central-1 does not respond allowing connection. Whereas an endpoint s3.eu-central-1.wasabisys.com does (for buckets in wasabi eu-central-1).

Any advice appreciated.
tetricky

**Comment by tetricky** [comment]
I am increasingly of the view that what I am seeing is a level of problem with networking, on my stack (may not have anything to do with longhorn specifically).

I am rebuilding it, and will report from there.  I would advise nobody to invest time or effort into this issue.

**Comment by joshimoo** [comment]
@tetricky thanks for letting us know, please feel free to open another issue if you are still having problems down the line.

**Comment by tetricky** [comment]
Just to confirm, on the off chance it's of some use to someone else. I have now rebuilt the stack, and all is now well.

Using Wasabi with my buckets in eu-europe-1; the endpoint that worked for me was: https://s3.eu-central-1.wasabisys.com

backupTarget: s3://[bucket name]@eu-central-1/[path within bucket]/
backupTargetCredentialSecret: wasabi-secret

-----------
apiVersion: v1
kind: Secret
metadata:
  name: wasabi-secret
  namespace: longhorn
type: Opaque
data:
  AWS_ACCESS_KEY_ID: [access key - base64 encoded]
  AWS_SECRET_ACCESS_KEY: [secret key base64 encoded]
  AWS_ENDPOINTS: [https://s3.eu-central-1.wasabisys.com base64 encoded]
-------------

**Comment by FredrickB** [comment]
> I am increasingly of the view that what I am seeing is a level of problem with networking, on my stack (may not have anything to do with longhorn specifically).
> 
> I am rebuilding it, and will report from there. I would advise nobody to invest time or effort into this issue.

Can confirm that a similar issue happens if DNS is not responding on the network during the installation and the backup storage is a remote S3 bucket hosted at a cloud provider referenced by domain name. I got the same error and realized my DNS server was down during the installation process. Fixing DNS and reinstalling resolved the issue. Leaving this here for anyone troubleshooting the same error message: Doublecheck DNS on the network before starting to troubleshoot Longhorn.

**Comment by LukasKnuth** [comment]
For future reference because I found this through Google:

1. Your secret needs to have the `AWS_ENDPOINTS` property set
2. It should be set to any URL from https://wasabi-support.zendesk.com/hc/en-us/articles/360015106031-What-are-the-service-URLs-for-Wasabi-s-different-storage-regions- **with** `https://` at the beginning
4. The "Backup Target" must be set to `s3://<bucket-name>@<region>/` (mind the trailing slash!)

**NOTE:** I wasn't able to get this working without using the region-specific S3 endpoint. Longhorn failed with the following error:

```
error listing backup volume names: failed to execute: /var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-v1.3.1/longhorn [backup ls --volume-only s3://test.longhorn-experiment@eu-central-1/], output failed to list objects with param: { Bucket: "test.longhorn-experiment", Delimiter: "/", Prefix: "/" } error: AWS Error: AuthorizationHeaderMalformed The authorization header is malformed; the region 'eu-central-1' is wrong; expecting 'us-east-1' <nil> 400 66C22CFC58C835E9 , stderr, time="2022-08-12T09:52:24Z" level=error
```
